tests/robustness/multiclient_test/framework/snapshotter.go

```diff
@@ -136,6 +143,31 @@ func (mcs *MultiClientSnapshotter) Cleanup() {
 	mcs.server.Cleanup()
 }
 
+// CleanupClient cleans up a given client. It delegates to the client's
+// Snapshotter Cleanup method, but also disconnects the client from the server,
+// removes the client from the server's user list, and removes the
+// ClientSnapshotter from MultiClientSnapshotter.
+func (mcs *MultiClientSnapshotter) CleanupClient(ctx context.Context) {
+	c := UnwrapContext(ctx)
+	if c == nil {
+		log.Println("Context does not contain a client")
+		return
+	}
+
+	mcs.mu.RLock()
+	s := mcs.clients[c.ID]
+	mcs.mu.RUnlock()
+
+	s.DisconnectClient(c.ID)
+	s.Cleanup()
+	mcs.server.RemoveClient(c.ID)
+
+	mcs.mu.Lock()
+	defer mcs.mu.Unlock()
+
+	delete(mcs.clients, c.ID)
```

Comment:
Suggestion: You probably don't need to use the lock twice: Use the W lock at 157 and obtain and delete the client immediately. Then clean up the client out of the mutex.

Reply:
Oh yeah, good point. There's no reason to separate the read and write in this case. Thanks!